README
Neolith - LPmud 伺服器
======================

Neolith 是一個修改自 MudOS 的 LPmud 系統，關於 LPmud 與 MudOS 的著作權聲明
以及 Neolith 和這二者的關係，請參考 doc/Copyright。 Neolith 中修改或新增的
任何程式碼及文件，除另有註明者外均依照自由軟體基金會所出版的 GPL version 2
公眾授權供所有人自由散佈使用。

 * 有關 GPL version 2 的授權內容請參考 COPYING。

 * 有關編譯與安裝 Neolith 的步驟請參考 INSTALL

 * 有關這個版本新增的功能與變動請參考 NEWS

 * 有關 Neolith 變動部份的詳細資料請參考 ChangeLog

 * 有關如何聯繫 Neolith 的作者請參考 AUTHORS


By Annihilator <annihilator@muds.net> (2001/05/01)
$Id: README,v 1.1.1.1 2002/11/23 07:57:08 annihilator Exp $


NEWS.md
※這個檔案列出的僅限於「使用者可見」的變動，有關所有修訂的詳細內容請參考
  ChangeLog。

Neolith-0.1.2 變動摘要
======================

 - 當 mudlib 不定義 master::creator_file() 時, 允許產生物件, 並將 uid
   設定為 NONAME。

 - 組態設定新增以下項目：

	StackSize		執行 LPC 程式的堆疊大小
	MaxLocalVariables	LPC 函式允許的區域變數個數
	MaxCallDepth		LPC 函式允許遞迴呼叫的深度
	ArgumentsInTrace	是否在 trace 訊息中列出函式參數值
	LocalVariablesInTrace	是否在 trace 訊息中列出區域變數值

   詳細說明請參考 neolith.conf 中的註解。

 - 預設標頭檔 (GlobalInclude) 不存在時, 不再產生編譯錯誤。

Neolith-0.1.1 變動摘要
======================

 - 不再允許使用 call_other 形式呼叫 static、private 或 protected 函數。

   理由如下:

   原 MudOS 允許在特定情況下使用 call_other 呼叫 static 與 private 函數(!?)
   ，這個情況是如果 call_other 的對象和 current object 相同。 (可能很多巫師
   不知道這一點) 允許這種呼叫會允許一種木馬攻擊方式, 在任何物件中:

   object ob;
   function f = (: call_other, $(ob), "any_private_function" :);

   如果攻擊者能夠將 f 成功地 bind 在 ob 上, 就可以從任何地方呼叫 ob 的私有
   函數, 而且 ob 的程式設計者完全沒有防止的辦法。

   在使用 function pointer 的 mudlib 中, 如果一個物件要「有條件」地允許其他
   物件呼叫自己的私有函數, 應該要「自己建立」這個 function pointer, 並且將
   之 bind 在它允許呼叫的物件上，也就是在 ob 的程式中必須明確寫有:

   function f = bind ((: some_private_function :), granted_object );

   才能將一個私有函數曝露給其他物件。

   封閉 call_other 呼叫 static 與 private 的函數功能讓私有函數可以真正的被
   私有，並且簡化 valid_bind 安全檢查的設計。

   其他影響: 某些使用函數名稱指定讓 driver 呼叫的函數, 如 sort_array, filter
   等傳入的函數名稱, 如果是 private 函數, 會失去效用, 請改成用 (: func :)
   形式, 這種形式速度會比原來使用字串的形式速度快一點。

 - efun time() 的傳回值改為在同一個控制流程 (control flow) 中不會改變。換句
   話說對 LPC 而言，不管機器 CPU 的速度快還是慢，在 LPC 將控制權交還給
   neolith 之前，time() 的傳回值都不會變。

   理由如下:

   原 MudOS 的 time() 會在每次被執行時呼叫真正的系統呼叫 time(2)，一般而言
   每個 LPC 控制流程執行時間都不會超過一秒鐘，但在某些情況下，例如用 LPC
   執行使用者存檔清查的迴圈，可能會用 set_eval_limit 強制執行需要較長時間
   的功能。在這種情形下，如果執行時間超過一秒鐘，time() 會傳回不同的時間。

   但是 MudOS 中其他還有許多和時間相關的 efun 如 query_idle, call_out, 等
   使用的是 backend 的 current_time 為計算基準，而傳統上 mudlib 中這些函
   數的時間會拿來和 time() 比較, time() 在用途上比較接近一個時間戳記
   (timestamp) 而非真正的時間，因此將 time() 的傳回值改為傳回 backend 的
   current_time，比較不會發生意料之外的情形。

   這同時也可以避免一些發生在秒和秒邊界的問題。

 - 新增支援 TELNET 通訊協定 RFC-1091 列編輯模式 (LINEMODE), 如果 client 端
   支援, 會打開本地列編輯模式 (即使用者輸入完一整列才送出一個封包), 使用列
   編輯模式可以大幅度降低封包數量, 讓頻寬利用更有效。

   telnet 使用者可以用 ^]display 測試自己的 telnet 程式是否支援 LINEMODE。

 - 完全移除 swap 機制。

   理由如下:

   以 ES2 實際測試結果, 300+ 人上線時, 約使用 15M 記憶體, swap 只移出不到
   300K, 效果不是很明顯。

   移除 swap 可以得到不少效能的提升, 並且在每個 object_t 少掉一個長整數
   swap_num (4 bytes), 在每個 program_t 少掉一個整數 line_swap_index
   (4 bytes) 的記憶體空間。

   300+ 人上線時 object 約 10000 個; program 約 2000 個, 移除 swap 共省下
   約 50K。

   以約 250K 的記憶體換提高的效率, 划算。

 - 以下 efun 新增不傳入任何參數的呼叫方式，此時會以 this_object() 為預設值:

   deep_inventory()
   environment()
   living()
   query_snoop()
   query_snooping()
   remove_interactive()
   userp()
   virtualp()

 - 新增使用者指令後面多餘的空白字元 (' ', \t, \v, \r, \f, \n) 在傳送給指
   令處理函式之前會自動被刪除掉。 (MudOS 只會刪除 ' ')

 - 變更 master object 的 get_root_uid() 與 get_bb_uid() 為選擇性提供, 如
   果 mudlib 不提供這兩個函式, neolith 仍可執行, 但 master object 與
   backbone 的 uid 將會是 NONAME。

 - 改良 cp() 檢查讀寫權的方式，現在 cp() 會先對第一個檔名呼叫 valid_read，
   若通過才會對第二個檔名呼叫 valid_write。

 - 刪除以下 Efun:

   terminal_colour

 - 修正物件在 reset() 中對自己 call_other()，導致 clean_up() 永遠不會被呼
   叫的問題，現在 reset() 與 clean_up() 的呼叫時機和說明文件所描述的一致。

 - 變更傳入 master object 的所有 valid_* apply 的檔名，現在不再帶有副檔名
   了，如果您的 mudlib 依賴某些 include 檔中定義的檔名 (如 ES2 mudlib)，
   像是 USER_OB、MASTER_OB 之類的，您必須將這些定義中的副檔名刪去才能維持
   原來的功能。例如: /include/globals.h 中定義:

   #define MASTER_OB	"/adm/obj/master.c"

   應改為:

   #define MASTER_OB	"/adm/obj/master"

 - 變更 user applies 函式名稱 'terminal_type' 為 'set_terminal_type' 以及
   'window_size' 為 'set_window_size'。

Neolith-0.1.0 變動摘要
======================

 - 更新 runtime config 檔的語法，並新增若干個新的設定項目，請參考
   src/neolith.conf 的範例及其中的註解說明。另外請使用 LPC/ 目錄下的
   runtime_config.h 取代您 mudlib 裡舊版的這個檔案。

 - 重新安排原始碼目錄結構，將原 MudOS 的原始碼依照功能分為數類各自以獨立的
   子目錄存放。

$Id: NEWS,v 1.2 2002/11/25 11:10:55 annihilator Exp $


src/rc.c
/*  $Id: rc.c,v 1.2 2002/11/25 11:11:05 annihilator Exp $

    This program is a part of Neolith project distribution. The Neolith
    project is based on MudOS v22pre5 LPmud driver. Read doc/Copyright
    before you try to use, modify or distribute this program.

    For more information about Neolith project, please visit:

    http://www.es2.muds.net/neolith

    ORIGINAL AUTHOR
	erikkay@mit.edu

    MODIFIED BY
	[Jul  4, 1994] by robo
	[Mar 26, 1995] by Beek
	[Jun 13, 2001] by Annihilator <annihilator@muds.net>
*/

#ifdef	HAVE_CONFIG_H
#include <config.h>
#endif /* HAVE_CONFIG_H */

#ifdef	STDC_HEADERS
#include <ctype.h>
#include <errno.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#endif /* STDC_HEADERS */

#include <sys/types.h>
#include <sys/stat.h>
#include <unistd.h>
#include <fcntl.h>

#include "lib/wrapper.h"
#include "rc.h"
#include "main.h"
#include "simulate.h"

/* global declarations */

char *config_str[NUM_CONFIG_STRS];
int config_int[NUM_CONFIG_INTS];

void init_config (char *config_file);
int get_config_item (svalue_t * res, svalue_t * arg);

/* static declarations */

static int fatal_config_error = 0;

static char *read_config_malloc (const char *filename);
static char *scan_config (char *config, char *name, int required, char *def);
static int scan_config_i (char *config, char *name, int required, int def);
static int scan_config_b (char *config, char *name, int required, int def);

/* implementations */

static char *
read_config_malloc (const char *filename)
{
  struct stat st;
#ifdef	HAVE_GETLINE
  char *line = NULL;
  size_t sz_line = 0;
#else /* ! HAVE_GETLINE */
  char line[1024];
#endif /* ! HAVE_GETLINE */
  char *conf, *p, *s;
  int fd;
  FILE *f;

  if (-1 == (fd = open (filename, O_RDONLY)))
    return NULL;

  if (-1 == fstat (fd, &st) || NULL == (f = fdopen (fd, "r"))) {
    close (fd);
    return NULL;
  }

  conf = (char *) calloc (st.st_size + 1, sizeof (char));
  if (!conf) {
    close (f);
    return NULL;
  }

#ifdef	HAVE_GETLINE
  for (p = conf; -1 != getline (&line, &sz_line, f);)
#else /* ! HAVE_GETLINE */
  for (p = conf; NULL != fgets (line, sizeof (line) - 1, f);)
#endif /* ! HAVE_GETLINE */
  {
    for (s = line; isspace (*s); s++);	/* skip leading blanks */
    if (*s == '#' || *s == '\n')
      continue;			/* skip comments and empty lines */

    p = stpcpy (p, s);
  }
  fclose (f);
#ifdef HAVE_GETLINE
  if (line)
    free (line);
#endif
  *p++ = '\0';

  return (char *) realloc (conf, p - conf);
}


/*
 * If the required flag is 0, it will only give a warning if the line is
 * missing from the config file.  Otherwise, it will give an error and exit
 * if the line isn't there.

   required:
      1  : Must have
      0  : optional
      -1 : warn if missing

 */
static char *
scan_config (char *config, char *name, int required, char *def)
{
  int sz_name;
  char *term, *p;

  sz_name = strlen (name);

  for (term = config; term; term = strchr (term, '\n'))
    {
      if (*term == '\n')
	term++;
      if (strncasecmp (name, term, sz_name) || !isspace (term[sz_name]))
	continue;

      *term = '#';

      term += sz_name;
      while (isspace (*term))
	term++;

      if ((p = strchr (term, '\n'))) {
	while (isspace (*(p-1)))
	  p--;
	*p = '\0';
      }
      term = xstrdup (term);
      if (p)
	*p = '\n';
      return term;
    }

  if (required < 0)
    {
      debug_message (_("warnning: %s is missing, assuming %s\n"),
		     name, def ? def : "null");
      return def;
    }

  if (required > 0)
    {
      debug_message (_("error: %s is missing\n"), name);
      fatal_config_error++;
      return NULL;
    }

  return def;
}

static int
scan_config_i (char *config, char *name, int required, int def)
{
  char *term;
  int n;

  if (NULL == (term = scan_config (config, name, required, NULL)))
    return def;

  n = atoi (term);
  free (term);

  return n;
}

static int
scan_config_b (char *config, char *name, int required, int def)
{
  char *p;
  int result = def;

  while ((p = scan_config (config, name, required, NULL)))
    {
      if (!strcasecmp (p, "yes"))
	result = 1;
      else if (!strcasecmp (p, "no"))
	result = 0;
      else
	debug_message (_("warnning: %s must be 'Yes' or 'No' [got: %s]\n"), name, p);
      free (p);
      required = 0;
    }

  return result;
}

void
init_config (char *filename)
{
  int i;
  char *config, *p;

  for (i = 0; i < NUM_CONFIG_INTS; i++)
    config_int[i] = 0;
  for (i = 0; i < NUM_CONFIG_STRS; i++)
    config_str[i] = 0;

  if (NULL == (config = read_config_malloc (filename)))
    {
      perror (filename);
      if (errno == ENOENT)
	fprintf (stderr, _("Use -f option to specify a config file\n"));
      exit (EXIT_FAILURE);
    }

  CONFIG_STR (__LOG_DIR__) = scan_config (config, "LogDir", 0, NULL);
  CONFIG_STR (__DEBUG_LOG_FILE__) =
    scan_config (config, "DebugLogFile", 0, NULL);
  CONFIG_INT (__ENABLE_LOG_DATE__) =
    scan_config_b (config, "LogWithDate", 0, 0);

  CONFIG_STR (__MUD_LIB_DIR__) = scan_config (config, "MudlibDir", 1, NULL);
  if (NULL == CONFIG_STR (__LOG_DIR__))
    CONFIG_STR (__LOG_DIR__) = CONFIG_STR (__MUD_LIB_DIR__);

  CONFIG_STR (__MUD_NAME__) = scan_config (config, "MudName", 0, NULL);
  CONFIG_STR (__ADDR_SERVER_IP__) =
    scan_config (config, "AddrServerIP", 0, NULL);

  CONFIG_STR (__GLOBAL_INCLUDE_FILE__) =
    scan_config (config, "GlobalInclude", 0, NULL);
  CONFIG_STR (__BIN_DIR__) = NULL;
  CONFIG_STR (__INCLUDE_DIRS__) = scan_config (config, "IncludeDir", 0, NULL);
  CONFIG_STR (__SAVE_BINARIES_DIR__) = scan_config (config, "SaveBinaryDir", 0, NULL);
  CONFIG_STR (__MASTER_FILE__) = scan_config (config, "MasterFile", 1, NULL);
  CONFIG_STR (__SIMUL_EFUN_FILE__) =
    scan_config (config, "SimulEfunFile", 0, NULL);
  CONFIG_STR (__DEFAULT_ERROR_MESSAGE__) =
    scan_config (config, "DefaultErrorMsg", 0, NULL);
  CONFIG_STR (__DEFAULT_FAIL_MESSAGE__) =
    scan_config (config, "DefaultFailMsg", 0, NULL);

  for (i = 0; i < 5 && (p = scan_config (config, "Port", 0, NULL)); i++)
    {
      char *typ;

      external_port[i].port = strtoul (p, &typ, 0);
      if (!strcasecmp (typ, ":telnet"))
	external_port[i].kind = PORT_TELNET;
      else if (!strcasecmp (typ, ":binary"))
	external_port[i].kind = PORT_BINARY;
      else if (!strcasecmp (typ, ":ascii"))
	external_port[i].kind = PORT_ASCII;
      else if (*typ)
	{
	  debug_message (_
			 ("*Protocol of port %d is invalid, assuming TELNET\n"),
			 external_port[i].port);
	  external_port[i].kind = PORT_TELNET;
	}
      free (p);
    }
  if (i < 1)
    {
      debug_message (_("*****No user-accessible ports assigned\n"));
      fatal_config_error++;
    }
  CONFIG_INT (__MUD_PORT__) = external_port[0].port;

  CONFIG_INT (__ADDR_SERVER_PORT__) =
    scan_config_i (config, "AddrServerPort", 0, 0);
  CONFIG_INT (__TIME_TO_CLEAN_UP__) =
    scan_config_i (config, "CleanupDuration", 0, 600);
  CONFIG_INT (__TIME_TO_RESET__) =
    scan_config_i (config, "ResetDuration", 0, 1800);
  CONFIG_INT (__INHERIT_CHAIN_SIZE__) =
    scan_config_i (config, "MaxInheritDepth", 0, 30);
  CONFIG_INT (__MAX_EVAL_COST__) =
    scan_config_i (config, "MaxEvaluationCost", 0, 1000000);
  CONFIG_INT (__MAX_ARRAY_SIZE__) =
    scan_config_i (config, "MaxArraySize", 0, 15000);
  CONFIG_INT (__MAX_BUFFER_SIZE__) =
    scan_config_i (config, "MaxBufferSize", 0, 4000000);
  CONFIG_INT (__MAX_MAPPING_SIZE__) =
    scan_config_i (config, "MaxMappingSize", 0, 15000);
  CONFIG_INT (__MAX_STRING_LENGTH__) =
    scan_config_i (config, "MaxStringLength", 0, 200000);
  CONFIG_INT (__MAX_BITFIELD_BITS__) =
    scan_config_i (config, "MaxBitFieldBits", 0, 1200);

  CONFIG_INT (__MAX_BYTE_TRANSFER__) =
    scan_config_i (config, "MaxByteTransfer", 0, 10000);
  CONFIG_INT (__MAX_READ_FILE_SIZE__) =
    scan_config_i (config, "MaxReadFileSize", 0, 20000);
  CONFIG_INT (__SHARED_STRING_HASH_TABLE_SIZE__) =
    scan_config_i (config, "SharedStringHashSize", 0, 20011);
  CONFIG_INT (__OBJECT_HASH_TABLE_SIZE__) =
    scan_config_i (config, "ObjectHashSize", 0, 10007);
  CONFIG_INT (__ENABLE_CRASH_DROP_CORE__) =
    scan_config_b (config, "CrashDropCore", 0, 1);
  CONFIG_INT (__EVALUATOR_STACK_SIZE__) =
    scan_config_i (config, "StackSize", 0, 1000);
  CONFIG_INT (__MAX_LOCAL_VARIABLES__) =
    scan_config_i (config, "MaxLocalVariables", 0, 25);
  CONFIG_INT (__MAX_CALL_DEPTH__) =
    scan_config_i (config, "MaxCallDepth", 0, 50);

  if (scan_config_b (config, "ArgumentsInTrace", 0, 0))
    g_trace_flag |= DUMP_WITH_ARGS;

  if (scan_config_b (config, "LocalVariablesInTrace", 0, 0))
    g_trace_flag |= DUMP_WITH_LOCALVARS;

  /*
   * from options.h
   */
  CONFIG_INT (__COMPILER_STACK_SIZE__) = 200;	/* CFG_COMPILER_STACK_SIZE; */
  CONFIG_INT (__LIVING_HASH_TABLE_SIZE__) = 256;	/* CFG_LIVING_HASH_SIZE; */

  free (config);

  if (fatal_config_error)
    {
      debug_message (_("*****Failed loading config (%d error%s): %s\n"),
		     fatal_config_error, fatal_config_error > 1 ? "s":"",
		     filename);
      exit (EXIT_FAILURE);
    }
}

int
get_config_item (svalue_t * res, svalue_t * arg)
{
  int num;

  num = arg->u.number;

  if (num < 0 || num >= RUNTIME_CONFIG_NEXT)
    return 0;

  if (num >= BASE_CONFIG_INT)
    {
      res->type = T_NUMBER;
      res->u.number = config_int[num - BASE_CONFIG_INT];
    }
  else
    {
      res->type = T_STRING;
      res->subtype = STRING_CONSTANT;
      res->u.string = config_str[num];
    }

  return 1;
}
